plugins/ingest/src/main/java/org/elasticsearch/ingest/processor/CompoundProcessor.java

```diff
@@ -59,23 +62,32 @@ public String getType() {
 
     @Override
     public void execute(IngestDocument ingestDocument) throws Exception {
-        try {
-            for (Processor processor : processors) {
+        for (Processor processor : processors) {
+            try {
                 processor.execute(ingestDocument);
-            }
-        } catch (Exception e) {
-            if (onFailureProcessors.isEmpty()) {
-                throw e;
-            } else {
-                executeOnFailure(ingestDocument);
+            } catch (Exception e) {
+                if (onFailureProcessors.isEmpty()) {
+                    throw e;
+                } else {
+                    executeOnFailure(ingestDocument, e, processor.getType());
+                }
+                return;
             }
         }
+    }
+
+    private void resetErrorContext(Map<String, String> ingestMetadata) {
 
     }
 
-    void executeOnFailure(IngestDocument ingestDocument) throws Exception {
+    void executeOnFailure(IngestDocument ingestDocument, Exception cause, String failedProcessorType) throws Exception {
+        Map<String, String> ingestMetadata = ingestDocument.getIngestMetadata();
+        ingestMetadata.put(ON_FAILURE_MESSAGE_FIELD, cause.getMessage());
+        ingestMetadata.put(ON_FAILURE_PROCESSOR_FIELD, failedProcessorType);
```

Comment:
yes, we can improve this in a follow up issue. I do think we should change metadata to be a map of maps and then get the entire exception into it (including all cause exceptions).